src/main/java/com/google/sps/servlets/SavedRecipeServlet.java

```diff
@@ -0,0 +1,131 @@
+    PreparedQuery results = datastore.prepare(query);
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      responseMap.put("recipeIsSaved", false);
+    } else {
+      String title = (String) entity.getProperty("title");
+      String imageUrl = (String) entity.getProperty("imageUrl");
+      String sourceUrl = (String) entity.getProperty("sourceUrl");
+      ArrayList<String> dietaryNeedsStrings = (ArrayList<String>) entity.getProperty("dietaryNeeds");
+      
+      // convert string to Diet enum because datastore stores dietaryNeeds as a list of strings
+      ArrayList<SavedRecipe.Diet> dietaryNeeds = new ArrayList<>();
+      for (String dietString: dietaryNeedsStrings) {
+        switch(dietString) {
+          case "VEGETARIAN":
+            dietaryNeeds.add(SavedRecipe.Diet.VEGETARIAN);
+            break;
+          case "VEGAN":
+            dietaryNeeds.add(SavedRecipe.Diet.VEGAN);
+            break;
+          case "GLUTENFREE":
+            dietaryNeeds.add(SavedRecipe.Diet.GLUTENFREE);
+            break;
+          case "DAIRYFREE":
+            dietaryNeeds.add(SavedRecipe.Diet.DAIRYFREE);
+            break;
+          default: break;
+        }
+      }
+      SavedRecipe savedRecipeObject = SavedRecipe.builder().setId(recipeId).setTitle(title).
+        setImage(imageUrl).setSourceUrl(sourceUrl).setDietaryNeeds(dietaryNeeds).build();
+          
+      responseMap.put("savedRecipe", savedRecipeObject);
+      responseMap.put("recipeIsSaved", true);
+    }
+    
+    Gson gson = new Gson();   
+    String json = gson.toJson(responseMap);
+    response.setContentType("application");
+    response.getWriter().println(json);
+  }
+
+  /** saves recipe information in datastore */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    response.setContentType("application/json");
+    JSONObject responseMap = new JSONObject();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    String recipeIdStr = request.getParameter("recipe-id");
+    Entity entity = new Entity("SavedRecipe", recipeIdStr);
+    
+    entity.setProperty("recipeId", Long.parseLong(recipeIdStr));
+
+    String title = request.getParameter("recipe-title");
+    entity.setProperty("title", title);
+
+    String imageUrl = request.getParameter("image-url");
+    entity.setProperty("imageUrl", imageUrl);
+
+    String sourceUrl = request.getParameter("source-url");
+    entity.setProperty("sourceUrl", sourceUrl);
+
+    String[] dietaryNeeds = request.getParameterValues("dietary-needs");
+    entity.setProperty("dietaryNeeds", Arrays.asList(dietaryNeeds));
+    
+    datastore.put(entity);
+
+    Gson gson = new Gson();   
+    String json = gson.toJson(responseMap);
+    response.setContentType("application");
+    response.getWriter().println(json);
```

Comment:
do you need the print statement? I don't see you using the return value for the post

Reply:
I may be missing this but are these saved recipes stored along with userid so that that specific user's saved recipes are stored togther

Reply:
joey <- That print statment was for testing.

coco <- I saved the recipes without the userid because the information about the recipe does not change from user to user so therefore it should be fine to have users to able to access the same recipe entity.